rust-runtime/aws-smithy-runtime/src/client/http/body/minimum_throughput.rs

```diff
@@ -0,0 +1,230 @@
+            let now = self.time_source.now();
+            let time_elapsed_since_earliest_poll = now
+                .duration_since(*earliest_time)
+                .map_err(|err| Error::TimeTravel(err.into()))?;
+
+            // This check ensures we that the data we're looking at covers a good range of time.
+            // If not, then we don't calculate a throughput.
+            if time_elapsed_since_earliest_poll < self.minimum_throughput.per_time_elapsed {
+                return Ok(None);
+            }
+
+            let total_bytes_logged =
+                self.throughput_logs
+                    .iter()
+                    .fold(0, |acc, (_, bytes_read)| acc + bytes_read) as f64;
+
+            Ok(Some(Throughput {
+                bytes_read: total_bytes_logged,
+                per_time_elapsed: time_elapsed_since_earliest_poll,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl<T: http_body::Body> MinimumThroughputBody<T> {
+    /// Given an HTTP body and a minimum throughput, create a new `MinimumThroughputBody`.
+    pub fn new(
+        time_source: impl TimeSource + 'static,
+        async_sleep: impl AsyncSleep + 'static,
+        body: T,
+        minimum_throughput: (u64, Duration),
+    ) -> Self {
+        Self {
+            inner: body,
+            throughput_logs: VecDeque::with_capacity(LOG_WINDOW_SIZE),
+            minimum_throughput: minimum_throughput.into(),
+            time_source: time_source.into_shared(),
+            async_sleep: async_sleep.into_shared(),
+        }
+    }
+}
+
+impl<T> http_body::Body for MinimumThroughputBody<T>
+where
+    T: http_body::Body<Data = bytes::Bytes, Error = Box<dyn std::error::Error + Send + Sync>>,
+{
+    type Data = T::Data;
+    type Error = T::Error;
+
+    fn poll_data(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        // It's possible that this body won't be polled again, so we ensure it'll
+        // wake by spawning a task with a reference to the waker, forcing another
+        // poll.
+        let wake_after = self.minimum_throughput.per_time_elapsed;
+        let waker = cx.waker().clone();
+        let sleeper = self.async_sleep.clone();
+        let _task = tokio::task::spawn(async move {
+            sleeper.sleep(wake_after).await;
+            waker.wake();
+        });
+        let this = self.as_mut().project();
+        let poll_res = this.inner.poll_data(cx);
+
+        let bytes_read = if let Poll::Ready(Some(Ok(ref data))) = poll_res {
+            data.remaining() as u64
+        } else {
+            0
+        };
+
+        let now = this.time_source.now();
+        this.throughput_logs.push_back((now, bytes_read));
+        // When the number of logs exceeds the window size, toss the oldest log.
+        if this.throughput_logs.len() > LOG_WINDOW_SIZE {
+            this.throughput_logs.pop_front();
+        }
```

Comment:
Seems like it should pop anything that's older than our time window, with the window size being a large upper bound. 16 is probably not high enough? Need to experiment with that number.